File: src/sync/webhook.py
"""Pub/Sub webhook handler for Gmail push notifications."""

from __future__ import annotations

import base64
import json
import logging

from src.db.connection import Database
from src.db.models import JobRepository, UserRepository

logger = logging.getLogger(__name__)


class WebhookHandler:
    """Handles Gmail Pub/Sub push notifications."""

    def __init__(self, db: Database):
        self.db = db
        self.users = UserRepository(db)
        self.jobs = JobRepository(db)

    def handle_notification(self, body: dict) -> bool:
        """Process a Pub/Sub push notification.

        Notification format:
        {
          "message": {
            "data": base64({emailAddress: "user@org.com", historyId: 12345}),
            "messageId": "...",
            "publishTime": "..."
          },
          "subscription": "projects/.../subscriptions/..."
        }
        """
        try:
            message = body.get("message", {})
            data_b64 = message.get("data", "")
            if not data_b64:
                logger.warning("Empty notification data")
                return False

            data = json.loads(base64.b64decode(data_b64))
            email_address = data.get("emailAddress", "")
            history_id = str(data.get("historyId", ""))

            if not email_address:
                logger.warning("No emailAddress in notification")
                return False

            # Look up user
            user = self.users.get_by_email(email_address)
            if not user:
                logger.warning("Unknown user in notification: %s", email_address)
                return False

            # Queue sync job
            self.jobs.enqueue("sync", user.id, {"history_id": history_id})
            logger.info("Queued sync for %s (history_id=%s)", email_address, history_id)
            return True

        except Exception as e:
            logger.error("Failed to process notification: %s", e)
            return False

File: src/api/routes/webhook.ts
import { Hono } from "hono";
import { getDb, schema } from "../../db/index.js";
import { getJobQueue } from "../../jobs/index.js";
import { eq } from "drizzle-orm";

export const webhookRoutes = new Hono();
const db = getDb();
const queue = getJobQueue();

// POST /webhook/gmail - Receive Gmail Pub/Sub push notifications
webhookRoutes.post("/gmail", async (c) => {
  try {
    const body = await c.req.json();

    // Extract and decode the notification
    const message = body.message;
    if (!message || !message.data) {
      return c.text("Invalid notification format", 400);
    }

    // Decode base64 data
    const decodedData = Buffer.from(message.data, "base64").toString("utf-8");
    const data = JSON.parse(decodedData);

    const { emailAddress, historyId } = data;

    // Look up user by email
    const user = await db
      .select()
      .from(schema.users)
      .where(eq(schema.users.email, emailAddress))
      .limit(1);

    if (user.length === 0) {
      console.warn(`Webhook notification for unknown user: ${emailAddress}`);
      return c.text("User not found", 200); // Return 200 to avoid retries
    }

    // Enqueue sync job with history ID
    await queue.enqueue("sync", user[0].id, {
      user_id: user[0].id,
      history_id: historyId,
      force_full: false,
    });

    console.log(`Webhook: Enqueued sync for user ${user[0].id} (history ${historyId})`);
    return c.text("OK", 200);
  } catch (error) {
    console.error("Webhook processing error:", error);
    return c.text("Internal processing error", 500);
  }
});
